---
layout: page
title: Running the pipeline
navigation: 3
---

# Running the pipeline
The pipeline is composed of three different modules within three different folders:

1. **NanoPreprocess** containing the preprocessing steps that will extract the data from raw fast5 files
1. **NanoTail** that will use the results produced by NanoPreprocess for estimating polyA sizes.  
1. **NanoMod** that will use the results produced by NanoPreprocess for estimating checmically modified sites.

Each module relies on tools installed within linux containers which recipes are stored within the **docker** folder. In case the *singularity* engine is chosen a number of images will be automatically generated and placed within the **singularity** folder.

## NanoPreprocess

Input files are either multifast5 or single fast5 files containing reads from direct RNA sequencing. 
They will be basecalled and eventually demultiplexed and aligned to a reference sequence (genome or transcriptome).

## Steps
 1. **testInput** Detection of kind of fast5 (multi or single)
 1. **baseCalling** Basecalling with *Albacore* or *Guppy*
 1. **demultiplexing_with_deeplexicon** Demultiplexing (optional) with **DeePlexiCon**
 1. **concatenateFastQFiles** This process concatenates the fastq files produces for each single basecalling 
 1. **QC** performed with  *MinIONQC*
 1. **fastQC** on fastq files
 1. **mapping** to the genome / transcriptome with either *minimap2* or *graphmap2*
 1. **counting** counts per gene, if mapping to the genome with *htseq-count*, or per transcript if mapping to the transcriptome with *NanoCount*. Reads are also assigned to a gene or to a transcript if they are uniquely mapping. A report file is also generated
 1. **alnQC** QC of aligned reads with *bam2stats*.
 1. **joinCountQCs** This process is for merging the report files generated by the counting step.
 1. **joinAlnQCs** This process is for merging the QC files generated by the alnQC step.
 1. **alnQC2** QC of aligned reads with *NanoPlot*. The plots PercentIdentityvsAverageBaseQuality_kde, LengthvsQualityScatterPlot_dot, HistogramReadlength and Weighted_HistogramReadlength are then merged together in a single picture.
 1. **multiQC**. Final repor enventually sent by mail too.
  

You can launch the pipeline choosing either the parameter **-with-singularity** or **-with-docker** depending on which containers you want to use:

## Input files

```
cd master_of_pores/NanoPreprocess

nextflow run nanopreprocess.nf --help
N E X T F L O W  ~  version 19.10.0
Launching `nanopreprocess.nf` [thirsty_bassi] - revision: a857b134a2
╔╦╗┌─┐┌─┐┌┬┐┌─┐┬─┐  ┌─┐┌─┐  ╔═╗╔═╗╦═╗╔═╗╔═╗
║║║├─┤└─┐ │ ├┤ ├┬┘  │ │├┤   ╠═╝║ ║╠╦╝║╣ ╚═╗
╩ ╩┴ ┴└─┘ ┴ └─┘┴└─  └─┘└    ╩  ╚═╝╩╚═╚═╝╚═╝
                                                                                       
====================================================
BIOCORE@CRG Preprocessing of Nanopore direct RNA - N F  ~  version 0.1
====================================================

kit                       : SQK-RNA002
flowcell                  : FLO-MIN106
fast5                     : ../real_data/test1wt/*.fast5
reference                 : ../anno/genome.fa.gz
annotation                : ../anno/Saccharomyces_cerevisiae.R64-1-1.98.gtf.gz

ref_type                  : genome
seq_type                  : RNA

output                    : test1
qualityqc                 : 5
granularity               : 

basecaller                : guppy
basecaller_opt            : 
GPU                       : OFF
demultiplexing            :  
demultiplexing_opt        : -m pAmps-final-actrun_newdata_nanopore_UResNet20v2_model.030.h5 

filter                    : 
filter_opt                : 
mapper                    : minimap2
mapper_opt                : -uf -k14
map_type                  : spliced

count                     : YES
counter_opt               : 

email                     :

```

You can change them by editing the **params.config** file or using the command line (each param name needs to have the characters **--** before): 

```bash
nextflow run nanopreprocess.nf -with-singularity -bg --output test2 > log.txt
```

To resume a previous execution that failed at a certain step or if you change a parameter that affects only some steps you can use the **Netxtlow** parameter **-resume** (only one dash!):


```bash
nextflow run nanopreprocess.nf -with-singularity -bg -resume > log.txt

...

[warm up] executor > crg
[e8/2e64bd] Cached process > baseCalling (RNA081120181_1)
[b2/21f680] Cached process > QC (RNA081120181_1)
[c8/3f5d17] Cached process > mapping (RNA081120181_1)
...

```

### Results:

-----------------------------------------------------


## NanoTail
Data produced by NanoPreprocess are needed for this module. 

Steps:
 1. **check_reference** It verifies whether the reference is zipped and eventually unzip it
 1. **tailfindr** it runs *tailfindr* tool in parallel.
 1. **collect_tailfindr_results** It collects the results of tailfindr.
 1. **filter_bam** Bam files are filtered with *samtools* to keep only mapped reads and remove secondary alignments
 1. **tail_nanopolish** It runs *nanopolish* in parallel.
 1. **collect_nanopolish_results** It collects the results of tail_nanopolish. 
 1. **join_results** It merges the results from the two algorithms and make a plot of the correlation.

....

### Results
Three folders contains the results:
1. NanoPolish: contains the output of *nanopolish* tool.
1. Tailfindr: contains the output of *tailfindr* tool.
1. PolyA_final: contains the txt files with the combined results (i.e. predicted polyA sizes). Here an example of a test:

```bash
"Read name"	"Tailfindr"	"Nanopolish"	"Gene Name"
"013a5dde-9c52-4de1-83eb-db70fb2cd130"	52.16	49.39	"YKR072C"
"01119f62-ca68-458d-aa1f-cf8c8c04cd3b"	231.64	274.28	"YDR133C"
"0154ce9c-fe6b-4ebc-bbb1-517fdc524207"	24.05	24.24	"YFL044C"
"020cde28-970d-4710-90a5-977e4b4bbc46"	41.27	56.79	"YGL238W"
```


## NanoMod
Data produced by NanoPreprocess are needed for this module. 

Finish writing usage

